### DISCUSSION TITLE
Rotate the EBSD data according to misorientation

### DISCUSSION THREAD

**Discussion by saberrider24** [mtex-toolbox/mtex#1928] [Ask Anything]
Hi,  everyone,

I made two EBSD scan on the same sample on the samle area, but the position of the sample is changed a bit. Now my aim is to rotate one of them so that they have the same orientation after the rotation. 

At first I wanted to use the rotation.byAxisAngle commend to rotate the data on the sample coordinate system, but I couldn't find a way to obtain the XYZ axis rotation angle based on the EBSD data. (If this is possible, I am glad to know how.)

So I am trying to rotate the EBSD data based on the misorientation between the same grain of two measurement.
```matlab
% grain 1 from the 1st scan
GM30=grainsM30(35).meanOrientation
% grain 1 from the 2st scan
GA79=grainsA79R(47).meanOrientation
%misorientation
Mis0 =  inv(GM30) * GA79
```
Here is the out come. 
```
Mis0 = misorientation (Ti3Al - alpha2 -> Ti Alpha 2)
 
  Bunge Euler angles in degree
     phi1     Phi    phi2    Inv.
  83.4657 131.816  234.86       0
```
Then I use this commend: 
```matlab
GM30R35=grainsM30(35).meanOrientation*rotation.byEuler(84*degree,132*degree,235*degree)
```
 to verify if the rotation is correct.

Here is the out come:
```
GM30R35 = orientation (Ti Alpha 2 -> xyz)
   Bunge Euler angles in degree
     phi1     Phi    phi2    Inv.
  132.871 102.965 189.776       0
```
The target grain orientation is this:
```
GA79 = orientation (Ti3Al - alpha2 -> xyz)
   Bunge Euler angles in degree
     phi1     Phi    phi2    Inv.
  132.717 102.554 189.959       0
```
They are almost identical. So I assume that the rotation is correct.

However, when I use the commend: 
```matlab
ebsdM30R=rotate(ebsdM30,rotation.byEuler(84*degree,132*degree,235*degree),'keepXY')
```
to rotate the whole EBSD data set, the grain(35)'s mean orientations became this:
```
grainsM30R(35).meanOrientation
ans = orientation (Ti Alpha 2 -> xyz)
 
  Bunge Euler angles in degree
     phi1     Phi    phi2    Inv.
  110.702 89.6852 316.679       0
```
The grain ID remain the same after the rotation, I verifed it with the ploted map. But the orientation is different from the out come of the rotation of a single grain's orientation. 

I don't know what went wrong. I think I misunderstood the rotate.byEuler commend for the EBSD data set. But I don't know how to correct it. Or I used a really bad method to rotate the data, perphas there are better ways to do that.  Please take a look at the code I used, any advice is apprciated. 

Cheers

Jason

**Comment by ralfHielscher** [comment]
Hi Jason,

this is indeed a bit complicated. Some basics to understand first:
1. Orientations are subject to symmetries, i.e., you will not be able to find a single rotation that maps all Euler angles of your first map onto the Euler angles of the second map. This is not a big problem, as it is sufficient that the orientations after the rotation are symmetrically equivalent.
2. If you multiply an orientation with a rotation you can do so from the left and from the right. If you do it from the left, you rotate the specimen - this is what you want and what the EBSD/rotate command does. 
3. If you multiply from the right you rotate the crystal coordinate system - this is what you do if you consider misorientations.

Because 1. it is not possible to find the rotation by one pair of grains alone. You have two options:
In the following `oriA` and `oriB` are the orientations of the corresponding grain(s) 

 **A - make use of the fact that the rotation is about the z-Axis** 
```matlab
%compute all possible rotations
rot = oriB .* oriA.CS * inv(oriA); 

% take the rotation with axis most parallel to z
[~,ind] = min(angle(axis(rot),zvector,'antipodal')
rot = rot(ind);
```
**B - consider more then one pair of grains**
```matlab
% generate some possible rotations covering all possible rotations
S3G = equispacedSO3Grid(specimenSymmetry,'points',100);
bestFit = inf;

% for each testing rotation 
for k = 1:length(S3G)
  % compute the fit between the pairs of orientations with respect to the rotation at hand
  fit = mean(angle(S3G(k).*oriA,oriB));
  % remember rotation if it was a best fit
  if fit < bestFit
    bestFit = fit;
    rot = S3G(k);
  end
end
% now we have a rough idea of the rotation
% hence we can choose the symmetry operations in oriB the same as in oriA
oriB = oriB.project2FundamentalRegion(rot * oriA);
% after oriA and oriB have the same symmetry operations, the misrotation is computed straight forward
rot = mean(oriB .* inv(oriA),'robust');
```
I hope this helps,

Ralf.

**Reply by saberrider24** [reply]
Dear Ralf,

Thank you for the reply! I found the rotation to obtain the symmetrically equivalent orientaiton for most of the grains by method of exhaustion. I tried all the possible rotation combination along XYZ axis. Now, after the rotation, the misorientation angle of the same grain between the two maps are less than 5 degree. But this is a really stupid method. So I would prefer to use a more scientific way. 

Both options are inspiring. Now I am trying to understand the code first. 

In option A:
It is quite straight forward if it is only around the Z-Axis. However, after the exhaustion method, I found that it is not only around the Z-Axis. But the code is inspiring to me so I would like understand it.
For the code: rot = oriB .* oriA.CS * inv(oriA);
rot=oriB*inv(oriA) is multiply from the right, so it rotates the crystal coordinate system.
Then the oriA.CS in between is to calculate all the equivalent orientation of oriA.

Am I right?

In option B:
I assume that it is to calculate the 'average misorientation' between all the equivalent orientation for oriA and oriB then rotate the map based on this misorientation. 
For the code:  fit = mean(angle(S3G(k).*oriA,oriB));
I assume the mean function is with respect to this:
https://mtex-toolbox.github.io/embedding.mean.html
So it take 100 points in the grid space that has the equal orientation and multiply it to oriA, and take the mean along the dimesion of the oriB. 
And the if cycle makes sure the smallest angle among those 100 points between the oriA and oriB are stored as S3G(k)
Then in oriB = oriB.project2FundamentalRegion(rot * oriA); the function is with respect to this page:
https://mtex-toolbox.github.io/orientation.project2FundamentalRegion.html
I do not understand from here, would you mind make a bit more explaination?

Best regards

Jason

**Reply by ralfHielscher** [reply]
I added some explanations to my answer. I  hope they make the situation more clear.

Ralf.
